inc/sitemaps/class-sitemaps-cache.php

```diff
@@ -41,6 +47,190 @@ public function is_enabled() {
 	}
 
 	/**
+	 * Get the cache key for a certain type and page
+	 *
+	 * Example key format for type "post", page 1: wpseo_sitemap_post_1:akfw3e_23azBa
+	 *
+	 * @param null|string $type The type to get the key for. Null or self::SITEMAP_INDEX_TYPE for index cache.
+	 * @param int         $page The page of cache to get the key for.
+	 *
+	 * @return string The key where the cache is stored on.
+	 */
+	public static function get_storage_key( $type = null, $page = 1 ) {
+
+		// Using SITEMAP_INDEX_TYPE for sitemap index cache.
+		$type = is_null( $type ) ? self::SITEMAP_INDEX_TYPE : $type;
+
+		$global_cache_validator = self::get_validator();
+		$type_cache_validator   = self::get_validator( $type );
+
+		$prefix  = self::$storage_key_prefix;
+		$postfix = sprintf( '_%d:%s_%s', $page, $global_cache_validator, $type_cache_validator );
+
+		$type = self::get_safe_type( $type, $prefix, $postfix );
+
+		// Build key.
+		$full_key = $prefix . $type . $postfix;
+
+		return $full_key;
+	}
+
+	/**
+	 * If the type is over length make sure we compact it so we don't have any database problems
+	 *
+	 * When there are more 'extremely long' post types, changes are they have variations in either the start or ending.
+	 * Because of this, we cut out the excess in the middle which should result in less chance of collision.
+	 *
+	 * @param string $type    The type of sitemap to be used.
+	 * @param string $prefix  The part before the type in the cache key. Only the length is used.
+	 * @param string $postfix The part after the type in the cache key. Only the length is used.
+	 *
+	 * @return string The type with a safe length to use
+	 *
+	 * @throws OutOfRangeException When there is less than 15 characters of space for a key that is originally longer.
+	 */
+	public static function get_safe_type( $type, $prefix = '', $postfix = '' ) {
+		// Length of key should not be over 53.
+		$max_length = 53;
+		$max_length -= strlen( 'timeout_' );
```

Comment:
This is a constant length so you don't have to calculate this every time. You can just 'precompile' the length.